plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/pvc/CommonPVCStrategyWithEphemeralWorkspacesSupport.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2016-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package com.redhat.che.multitenant.pvc;
+
+import static org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesObjectUtil.newVolumeMount;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.provision.LogsVolumeMachineProvisioner.LOGS_VOLUME_NAME;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodSpec;
+import io.fabric8.kubernetes.api.model.VolumeBuilder;
+import java.util.Map;
+import java.util.Map.Entry;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.config.Volume;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.WorkspaceManager;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.CommonPVCStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.PVCSubPathHelper;
+
+/**
+ * Extends {@code CommonPVCStrategy} and provides support of the ephemeral workspaces based on the
+ * `emptyDir` volumes. If workspace config contains `mountSources` attribute which is set to true,
+ * than instead of using regular {@code CommonPVCStrategy}, each volume would be represented as an
+ * `emptyDir`. NOTE: no data would be persisted once between ephemeral workspace restarts, since
+ *
+ * @see <a href="https://kubernetes.io/docs/concepts/storage/volumes/#emptydir">EmptyDir</a>
+ * @author Ilya Buziuk
+ */
+public class CommonPVCStrategyWithEphemeralWorkspacesSupport extends CommonPVCStrategy {
```

Comment:
This class has pretty complex logic. Consider adding unit tests to avoid a situation when someone else breaking the prod after changes in a class he is not the author of.

Reply:
sure